omc3/optics_measurements/crdt.py

```diff
@@ -0,0 +1,237 @@
+    df.columns = [f'{col}_{suffix}' if col not in exceptions else col for col in df.columns]
+    return df
+
+
+def get_crdt_phases(joined_dfs, crdt, lines_and_phases, phase_sign):
+    phases, err_phases = np.zeros((len(joined_dfs), len(joined_dfs[0]))), np.zeros((len(joined_dfs), len(joined_dfs[0])))
+
+    # don't know the exact way to get to this via line indices so for now only via hacky way
+    signflip = 1
+    if crdt['term'] in ["F_NO2", "F_NO1"]:
+        signflip = -1
+    for idx, joined_df in enumerate(joined_dfs):
+        phases[idx, :] = (phase_sign*joined_df[lines_and_phases[PHASE]] -
+                          crdt['line'][0]*joined_df['MUX'] -
+                          crdt['line'][1]*joined_df['MUY'] -
+                          signflip*0.75 -
+                          0.5*((np.abs(crdt['line'][0])+np.abs(crdt['line'][1]))//3)
+                          ) % 1
+        try:
+            err_phases[idx, :] = np.sqrt(joined_df[lines_and_phases[f'{ERR}{PHASE}']]**2+
+                                         (crdt['line'][0]*joined_df['ERRMUX'])**2+
+                                         (crdt['line'][1]*joined_df['ERRMUY'])**2)
+        except KeyError:
+            err_phases[idx, :] = np.NaN
+
+    return circular_nanmean(phases, axis=0, errors=err_phases), circular_nanerror(phases, axis=0, errors=err_phases)
+
+
+def get_crdt_invariant(crdt, invariants):
+    exp = {'X': np.abs(crdt['line'][0]), 'Y': np.abs(crdt['line'][1])}
+    exp[crdt['plane']] = exp[crdt['plane']] - 1 # to compensate for the normalization with tune line
+
+    crdt_invariant = invariants['X'].T[0]**exp['X']*invariants['Y'].T[0]**exp['Y']
+    err_crdt_invariant = np.sqrt((exp['X']*crdt_invariant/invariants['X'].T[0])**2*
+                                 invariants['X'].T[1]**2 +
+                                 (exp['Y']*crdt_invariant/invariants['Y'].T[0])**2*
+                                 invariants['Y'].T[1]**2)
+    return crdt_invariant, err_crdt_invariant
+
+
+def get_crdt_amplitude(joined_dfs, crdt, invariants, lines_and_phases):
+    crdt_invariants, err_crdt_invariants = get_crdt_invariant(crdt, invariants)
+
+    assert len(crdt_invariants) == len(joined_dfs)
+
+    amps, err_amps = np.zeros(len(joined_dfs[0])), np.zeros(len(joined_dfs[0]))
+
+    for idx, bpm in enumerate(joined_dfs[0].index):
+        lineamplitudes, err_lineamplitudes = [], []
+        for joined_df in joined_dfs:
+            lineamplitudes.append(joined_df.loc[bpm, lines_and_phases[AMPLITUDE]])
+            try:
+                err_lineamplitudes.append(joined_df.loc[bpm, lines_and_phases[f'{ERR}{AMPLITUDE}']])
+            except KeyError:
+                err_lineamplitudes.append(0.)
+
+        amps[idx], err_amps[idx] = fit_amplitude(lineamplitudes,
+                                                 err_lineamplitudes,
+                                                 crdt_invariants,
+                                                 err_crdt_invariants)
```

Comment:
this should be vectorised, sorry that I did not notice the loop over BPMs before.